Generate cases improved (#106)
testomatio/skills: skills/generate-cases/references/writing-rule.md

## Test Suite Writing Rules

Formulas, business rules, edge-case reasoning, and feature

> context live in the suite/test description.

Reviewer: ?

Prerequisites which are common to all test cases must be written as bullet points in the suite description.
Formulas, diagrams, relavant to all test cases can be included as codeblocks in the suite description.

## Test Case Writing Rules

Describe the intent of the test case.

> Intent is needed only if title is not fully enough to explain the intent.

Reviewer: ?

Reviewer: intent should be description

Test case consist of description and steps
Description must be clear and concise.

> Focus on whitebox testing, thus each operation and observable results must be obtained via public apis or UI.

Reviewer: ?

Reviewer: blackbox

Reviewer: unless specified different

Reviewer: prefer FE tests (based on testomat test cases, FE repo)

Reviewer: skill should ask, not suggest

Reviewer: RECOMMEND plan mode

Prefer using UI over public apis when possible.
If you think UI is available, and test can be achived from UI, test must use it

> Understand UI pages and components or API endpoints from source code. If source code not availble ask user for it.

Author: ```suggestion
Understand UI pages and components or API endpoints from source code or design or requirements. 
```

Add system checks only if it not clear how to test via public apis or UI.
It is not a unit test, usually no direct server or code access is allowed.

> Why in description, what in steps. 

Reviewer: ?

Reviewer: "Why" in description, "what" in steps.

Put preparations into **preconditions**/**description** section, not as steps.

## Preconditions  

Do not include preconditions like "service is running" if it's not explicitly mentioned by user or is not actually what the test is about.
Define the pre-conditions and initial state of the system as bullet points.
Avoid if not needed or is the same as the suite description.
If relevant, include the user role

## Steps

Step should consist of **action** (with optional test data) and **expected result**.
Each step must be simple sentences.
Avoid using commas, subsentances
Steps are mechanical: click, send, read, assert.
Each step must include exact instructions

> Prefer steps that include specific values, urls

Reviewer: ?

Reviewer: do not specific values unless they are important for scenario (e.g. boundary values)

Reviewer: "url" mean specific **path**, not domain

Prefer concrete values over general words. Replace "small", "known", "around", "e.g.", "like", "(…)" with literal numbers, strings, etc. 
If no values are availble use placholders variable names like `${url}` or `${company-title}`
Avoid general statements in steps. Move general statements to the description.
Do not chain multiple distinct actions or use unnecessary And / Or combinations in a single sequence.

> All step actions must be clear to perform via PUBLIC API or UI

Reviewer: ?

> All checks/verifications/assertions should be in **expected result**, not as separate steps.

Reviewer: ?

Reviewer: not in step "action"

You may add a codeblock after a step if needed to:

- write API request
- write SQL
- write shell command
- to illustrate the point using pseudocode
- etc

> Avoid using bold/italic formatting in steps.

Reviewer: ? why

Reviewer: rules of applying bold, italic

If expected result has multiple conditions split them into separate lines:

Instead of:
  *Expected:* Comment appears in the Run's comment list with the current user as author*

Use:
  *Expected:* Comment appears in the Run's comment list
  *Expected:* Current user is the author

## AI Writing Patterns to Avoid

Avoid:

Puffery: pivotal, crucial, vital, testament, enduring legacy
Empty "-ing" phrases: ensuring reliability, showcasing features, highlighting capabilities
Promotional adjectives: groundbreaking, seamless, robust, cutting-edge
Overused AI vocabulary: delve, leverage, multifaceted, foster, realm, tapestry
Formatting overuse: excessive bullets, emoji decorations, bold on every other word

Be specific, not grandiose. Say what it actually does. Avoid marketing language.

> Words to watch: stands/serves as, is a testament/reminder, plays a vital/significant/crucial/pivotal role, underscores/highlights its importance/significance, reflects broader, symbolizing its ongoing/enduring/lasting impact, key turning point, indelible mark, deeply rooted, profound heritage, steadfast dedication...

Reviewer: don't understand the purpose of this
